### DISCUSSION TITLE
Cannot get dynamic dropdown element for Angular web application to be recognized in Cypress

### DISCUSSION THREAD

**Discussion by rnunezil** [cypress-io/cypress#24978] [Questions and Help]
I have a web application written in Angular I am trying to test in Cypress. I am having trouble making the dropdown menu show up(circled in attached image on left). The dropdown menu shows up when clicking on the 3 dots. When clicking on the dots, I notice the class attribute in the 'fui-datagrid-action-menu' changes to the class attribute show encircled. Thus in Cypress, I tried to set the class attribute value to that value encircled. But when running my test, I get a 'Expected to find element: fui-dropdown, but never found it.'. Below is a snippet of my Cypress code. Can anybody suggest what I am doing wrong or is there a better way of coding for a page like this where the Angular elements are dynamic(ie, class attributes change, etc). Thanks.

![image](https://user-images.githubusercontent.com/106498583/205722953-790a26f4-6817-4d93-bc65-269154ffb00a.png)
```
cy.get('fui-widget-body')
                    .within(() => {
                  cy.log('user found');
                  cy.get('fui-datagrid-action-menu').invoke('show').trigger('mouseover', {force: true});
                  cy.get('fui-datagrid-action-menu').invoke('show').click({ force: true });
                  cy.get('fui-datagrid-action-menu').invoke('attr', 'class', 'fui-datagrid-body-row-action-menu ' +
                    'ng-star-inserted fui-datagrid-action-menu-visible fui-datagrid-action-menu-open')
                    .within(($menu) => {
                      cy.get('fui-dropdown').invoke('show').click({ force: true });
```

**Comment by portyu9** [comment]
Hi @rnunezil 

The main problem is that you are trying to open the Angular component by manually changing its CSS class:

```js
.invoke('attr', 'class', ...)
```

That changes the DOM attribute, but it does **not** update Angular's component state or run the click/event handlers that normally open the menu. The same applies to repeatedly using `.invoke('show')` and `{ force: true }`: those can make DOM/CSS interactions happen, but they do not reproduce the application's actual state transition.

Instead, click the real three-dot trigger and let Angular open the dropdown:

```js
cy.get('fui-widget-body')
  .find('fui-datagrid-action-menu')
  .should('be.visible')
  .click()

cy.get('fui-dropdown')
  .should('exist')
  .and('be.visible')
  .contains('Assign role')
```

If the clickable three-dot element is a child of `fui-datagrid-action-menu`, target that child instead. Ideally add a stable selector such as `data-cy` to the trigger rather than depending on Angular-generated classes.

Also be careful with `.within()`. Every `cy.get()` inside it is scoped to that element. If the Angular dropdown is rendered through an overlay/portal elsewhere in the document, this will never find it:

```js
cy.get('fui-datagrid-action-menu').within(() => {
  cy.get('fui-dropdown')
})
```

In that case, click the trigger inside the row/menu and then start a new query from the document, as in the first example.

I would remove the class mutation and forced clicks first. Test the same interaction a user performs and assert that Angular creates/displays the dropdown afterward.

Please mark this answer as accepted if it helped, thank you.
